File: .pub-cache/hosted/pub.dartlang.org/flutter_socket_io-0.6.0/android/src/main/java/com/itsclicking/clickapp/fluttersocketio/SocketIOManager.java
package com.itsclicking.clickapp.fluttersocketio;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodChannel;

public class SocketIOManager implements ISocketIOManager {

    private static final String TAG = "SocketIOManager";
    private static ISocketIOManager mInstance;
    private Map<String, SocketIO> mSockets;


    public synchronized static ISocketIOManager getInstance() {
        if (mInstance == null) {
            mInstance = new SocketIOManager();
        }
        return mInstance;
    }

    private SocketIOManager() {
        mSockets = new HashMap<>();
    }

    private SocketIO getSocket(String socketId) {
        if(mSockets != null && !Utils.isNullOrEmpty(socketId)) {
            Utils.log("TOTAL SOCKETS: ", String.valueOf(mSockets.size()));
            return mSockets.get(socketId);
        } else {
            Utils.log("TOTAL SOCKETS: ", "NULL");
        }
        return null;
    }

    private boolean isExistedSocketIO(String socketId) {
        SocketIO socketIO = getSocket(socketId);
        return (socketIO != null);
    }

    private void addSocketIO(SocketIO socketIO) {
        if(mSockets == null) {
            mSockets = new HashMap<>();
        }
        if(!isExistedSocketIO(socketIO.getId())) {
            Utils.log("added SocketIO", socketIO.getId());
            mSockets.put(socketIO.getId(), socketIO);
        }
    }

    private void removeSocketIO(SocketIO socketIO) {
        if(mSockets != null) {
            mSockets.remove(socketIO.getId());
        }
    }

    private boolean isConnected(SocketIO socketIO) {
        return socketIO != null && socketIO.isConnected();
    }

    private String getSocketId(String domain, String namespace) {
        if(!Utils.isNullOrEmpty(domain)) {
            return domain + (namespace != null ? namespace : "");
        }
        return null;
    }

    private SocketIO createSocketIO(MethodChannel channel, String domain,
                                    String namespace, String query, String callback) {
        return new SocketIO(channel, domain, namespace, query, callback);
    }

    @Override
    public void init(MethodChannel channel, String domain, String namespace, String query, String callback) {
        if(isExistedSocketIO(getSocketId(domain, namespace))) {
            Utils.log(TAG, "socket: " + getSocketId(domain, namespace) + " already existed!");
        } else {
            SocketIO socketIO = createSocketIO(channel, domain, namespace, query, callback);
            addSocketIO(socketIO);
            socketIO.init();
        }
    }

    @Override
    public void connect(String domain, String namespace) {
        SocketIO socketIO = getSocket(getSocketId(domain, namespace));
        if(socketIO != null) {
            socketIO.connect();
        } else {
            Utils.log(TAG, "socket: " + getSocketId(domain, namespace) + " is not initialized!");
        }
    }

    @Override
    public void sendMessage(String domain, String namespace, String event, String message, String callback) {
        SocketIO socketIO = getSocket(getSocketId(domain, namespace));
        if(socketIO != null) {
            socketIO.sendMessage(event, message, callback);
        } else {
            Utils.log(TAG, " not found socket: " + getSocketId(domain, namespace));
        }
    }

    @Override
    public void subscribes(String domain, String namespace, Map<String, String> subscribes) {
        SocketIO socketIO = getSocket(getSocketId(domain, namespace));
        if(socketIO != null) {
            socketIO.subscribes(subscribes);
        } else {
            Utils.log(TAG, " not found socket: " + getSocketId(domain, namespace));
        }
    }

    @Override
    public void unSubscribes(String domain, String namespace, Map<String, String> subscribes) {
        SocketIO socketIO = getSocket(getSocketId(domain, namespace));
        if(socketIO != null) {
            socketIO.unSubscribes(subscribes);
        } else {
            Utils.log(TAG, " not found socket: " + getSocketId(domain, namespace));
        }
    }

    @Override
    public void unSubscribesAll(String domain, String namespace) {
        Utils.log(TAG, "--- START unSubscribesAll ---");
        SocketIO socketIO = getSocket(getSocketId(domain, namespace));
        if(socketIO != null) {
            socketIO.unSubscribesAll();
        } else {
            Utils.log(TAG, " not found socket: " + getSocketId(domain, namespace));
        }
        Utils.log(TAG, "--- END unSubscribesAll ---");
    }

    @Override
    public void disconnect(String domain, String namespace) {
        Utils.log(TAG, "--- START disconnect ---");
        SocketIO socketIO = getSocket(getSocketId(domain, namespace));
        if(socketIO != null) {
            socketIO.disconnect();
        } else {
            Utils.log(TAG, " not found socket: " + getSocketId(domain, namespace));
        }
        Utils.log(TAG, "--- END disconnect ---");
    }

    @Override
    public void destroySocket(String domain, String namespace) {
        if(!Utils.isNullOrEmpty(mSockets)) {
            SocketIO socketIO = getSocket(getSocketId(domain, namespace));
            if(socketIO != null) {
                removeSocketIO(socketIO);
                socketIO.destroy();
            }  else {
                Utils.log(TAG, " not found socket: " + getSocketId(domain, namespace));
            }
        }
    }

    @Override
    public void destroyAllSockets() {
        if(!Utils.isNullOrEmpty(mSockets)) {
            for(Map.Entry<String, SocketIO> item : mSockets.entrySet()) {
                if(item != null) {
                    SocketIO socket = item.getValue();
                    if(socket != null) {
                        socket.destroy();
                    }
                }
            }
            mSockets.clear();
        }
    }

    public static class MethodCallArgumentsName {
        public static final String SOCKET_DOMAIN = "socketDomain";
        public static final String SOCKET_NAME_SPACE = "socketNameSpace";
        public static final String SOCKET_CALLBACK = "socketCallback";
        public static final String SOCKET_EVENT = "socketEvent";
        public static final String SOCKET_MESSAGE = "socketMessage";
        public static final String SOCKET_DATA = "socketData";
        public static final String SOCKET_QUERY = "socketQuery";
    }

    public static class MethodCallName {
        public static final String SOCKET_INIT = "socketInit";
        public static final String SOCKET_CONNECT = "socketConnect";
        public static final String SOCKET_DISCONNECT = "socketDisconnect";
        public static final String SOCKET_SUBSCRIBES = "socketSubcribes";
        public static final String SOCKET_UNSUBSCRIBES = "socketUnsubcribes";
        public static final String SOCKET_UNSUBSCRIBES_ALL = "socketUnsubcribesAll";
        public static final String SOCKET_SEND_MESSAGE = "socketSendMessage";
        public static final String SOCKET_DESTROY = "socketDestroy";
        public static final String SOCKET_DESTROY_ALL = "socketDestroyAll";
    }
}

File: .pub-cache/hosted/pub.dartlang.org/flutter_socket_io-0.6.0/example/ios/Runner/SocketObj/SocketIOManager.m
//
//  SocketIOManager.m
//  SocketTest
//
//  Created by AnhNguyen on 8/18/18.
//  Copyright © 2018 ATA_Studio. All rights reserved.
//

#ifndef SYNTHESIZE_SINGLETON_FOR_CLASS
#define SYNTHESIZE_SINGLETON_FOR_CLASS(classname)           \
\
+ (classname *)shared##classname {                          \
static dispatch_once_t pred;                            \
static classname * shared##classname = nil;             \
dispatch_once( &pred, ^{                                \
shared##classname = [[self alloc] init];            \
});                                                     \
return shared##classname;                               \
}
#endif

#import "SocketIOManager.h"
#import "UtilsSocket.h"

@interface SocketIOManager ()

- (SocketIO *)getSocket:(NSString *)socketId;
- (BOOL)isExistedSocketIO:(NSString *)socketId;
- (void)addSocketIO:(SocketIO *)socketIO;
- (void)removeSocketIO:(SocketIO *)socketIO;
- (BOOL)isConnected:(SocketIO *)socketIO;
- (NSString *)getSocketId:(NSString *)domain nameSpace:(NSString *)nameSpace;
- (SocketIO *)createSocketIO:(FlutterMethodChannel *)channel
                      domain:(NSString *)domain
                       query:(NSDictionary *)query
                   nameSpace:(NSString *)nameSpace
              statusCallback:(NSString *)callBack;

@end

@implementation SocketIOManager

SYNTHESIZE_SINGLETON_FOR_CLASS(SocketIOManager)

- (SocketIO *)getSocket:(NSString *)socketId {
    SocketIO *result = nil;
    if (self.mSockets && ![UtilsSocket isNullOrEmpty:self.mSockets]) {
        NSLog(@"TOTAL SOCKETS: %zi", self.mSockets.count);
        result = [self.mSockets objectForKey:socketId];
    } else {
        NSLog(@"TOTAL SOCKETS: NULL");
    }
    return result;
}

- (BOOL)isExistedSocketIO:(NSString *)socketId {
    if (socketId) {
        if (![UtilsSocket isNullOrEmpty:self.mSockets]) {
            SocketIO *socketIO = [self getSocket:socketId];
            if (socketIO) {
                return YES;
            }
        }
    }
    return NO;
}

- (void)addSocketIO:(SocketIO *)socketIO {
    if (!self.mSockets) {
        self.mSockets = [[NSMutableDictionary alloc] init];
    }
    if (![self isExistedSocketIO:socketIO.getId]) {
        NSLog(@"ADDED SOCKETIO %@", socketIO.getId);
        [self.mSockets setValue:socketIO forKey:socketIO.getId];
    }
}

- (void)removeSocketIO:(SocketIO *)socketIO {
    if (self.mSockets) {
        [self.mSockets removeObjectForKey:socketIO.getId];
    }
}

- (BOOL)isConnected:(SocketIO *)socketIO {
    return socketIO && socketIO.isConnected;
}

- (NSString *)getSocketId:(NSString *)domain nameSpace:(NSString *)nameSpace {
    NSString *result = nil;
    if (domain) {
        if (nameSpace) {
            result = [NSString stringWithFormat:@"%@%@", domain, nameSpace];
        } else {
            result = domain;
        }
    }
    return result;
}

- (SocketIO *)createSocketIO:(FlutterMethodChannel *)channel
                      domain:(NSString *)domain
                       query:(NSDictionary *)query
                   nameSpace:(NSString *)nameSpace
              statusCallback:(NSString *)callBack {
    return [SocketIO initSocketIO:channel query:query domain:domain nameSpace:nameSpace statusCallback:callBack];
}


#pragma mark SocketIO

- (void)initSocket:(FlutterMethodChannel *)channel
            domain:(NSString *)domain
             query:(id)query
          namspace:(NSString *)namspace
    callBackStatus:(NSString *)callback {
    if ([self isExistedSocketIO:[self getSocketId:domain nameSpace:namspace]]) {
        NSLog(@"SOCKET %@ ALREADY EXISTED!", [self getSocketId:domain nameSpace:namspace]);
    } else {
        SocketIO *socketIO = [self createSocketIO:channel domain:domain query:query nameSpace:namspace statusCallback:callback];
        [self addSocketIO:socketIO];
        [socketIO initSocket];
    }
}

- (void)connectSocket:(NSString *)domain
             namspace:(NSString *)namspace {
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [socketIO connect];
    } else {
        NSLog(@"SOCKET %@ IS NOT INITIALIZED", [self getSocketId:domain nameSpace:namspace]);
    }
}

- (void)sendMessage:(NSString *)event
            message:(NSString *)message
             domain:(NSString *)domain
           namspace:(NSString *)namspace
     callBackStatus:(NSString *)callback {
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [socketIO sendMessage:event mess:message callBack:callback];
    } else {
        [self errorNotFoundSocket:domain namspace:namspace];
    }
}

- (void)subscribes:(NSString *)domain
          namspace:(NSString *)namspace
        subscribes:(NSMutableDictionary *)subscribes {
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [socketIO subscribeList:subscribes];
    } else {
        [self errorNotFoundSocket:domain namspace:namspace];
    }
}

- (void)unSubscribes:(NSString *)domain
            namspace:(NSString *)namspace
          subscribes:(NSMutableDictionary *)subscribes {
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [socketIO unSubscribeList:subscribes];
    } else {
        [self errorNotFoundSocket:domain namspace:namspace];
    }
}

- (void)unSubscribesAll:(NSString *)domain
               namspace:(NSString *)namspace {
    NSLog(@"----- START UNSUBSCRIBESALL -----");
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [socketIO unSubscribeAll];
    } else {
        [self errorNotFoundSocket:domain namspace:namspace];
    }
    NSLog(@"----- END UNSUBSCRIBESALL -----");
}

- (void)disconnectDomain:(NSString *)domain
                namspace:(NSString *)namspace {
    NSLog(@"----- START DISCONNECT -----");
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [socketIO disconnect];
    } else {
        [self errorNotFoundSocket:domain namspace:namspace];
    }
    NSLog(@"----- END DISCONNECT -----");
}

- (void)destroySocketDomain:(NSString *)domain
                   namspace:(NSString *)namspace {
    SocketIO *socketIO = [self getSocket:[self getSocketId:domain nameSpace:namspace]];
    if (socketIO) {
        [self removeSocketIO:socketIO];
        [socketIO destroy];
    } else {
        [self errorNotFoundSocket:domain namspace:namspace];
    }
}

- (void)destroyAllSocket {
    if ([UtilsSocket isNullOrEmpty:self.mSockets]) {
        NSArray *keys = [self.mSockets allKeys];
        for (id key in keys) {
            SocketIO *socket = [self.mSockets objectForKey:key];
            if (socket) {
                [socket destroy];
            }
        }
        [self.mSockets removeAllObjects];
    }
}

- (void)errorNotFoundSocket:(NSString *)domain
                   namspace:(NSString *)namspace {
    NSLog(@"NOT FOUND SOCKET : %@", [self getSocketId:domain nameSpace:namspace]);
}

@end
